File: BOJ/♣ 14000~14999/14888_연산자_끼워넣기.cpp
#include <iostream>
using namespace std;

int n;
int num_list[12];
int max_ = -2100000000;
int min_ = 2100000000;

void backtracking(int value, int count, int plus, int minus, int mul, int div)
{
    if (count == n)
    {
        if (value > max_)
            max_ = value;
        if (value < min_)
            min_ = value;
        return;
    }

    if (plus > 0)
        backtracking(value + num_list[count + 1], count + 1, plus - 1, minus, mul, div);
    if (minus > 0)
        backtracking(value - num_list[count + 1], count + 1, plus, minus - 1, mul, div);
    if (mul > 0)
        backtracking(value * num_list[count + 1], count + 1, plus, minus, mul - 1, div);
    if (div > 0)
        backtracking(value / num_list[count + 1], count + 1, plus, minus, mul, div - 1);
}

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(NULL);
    cout.tie(NULL);

    cin >> n;
    for (int i = 1; i <= n; i++)
        cin >> num_list[i];
    int a, b, c, d;
    cin >> a >> b >> c >> d;
    backtracking(num_list[1], 1, a, b, c, d);

    cout << max_ << '\n'
         << min_;
}

File: BOJ/♣ 14000~14999/14888_연산자_끼워넣기.py
import itertools


def calculate(operation_order):
    result = A[0]

    for i in range(N - 1):
        operator = operation_order[i]

        if operator == "+":
            result += A[i + 1]
        elif operator == "-":
            result -= A[i + 1]
        elif operator == "*":
            result *= A[i + 1]
        elif result >= 0:
            result //= A[i + 1]
        else:
            result = -result // A[i + 1] * -1

    return result


N = int(input())
A = list(map(int, input().split()))
plus, minus, mul, div = list(map(int, input().split()))
operators = ["+"] * plus + ["-"] * minus + ["*"] * mul + ["//"] * div
permutations = itertools.permutations(operators)

max_value = -(10**18)
min_value = 10**18

for operation_order in permutations:
    cur_value = calculate(operation_order)
    max_value = max(max_value, cur_value)
    min_value = min(min_value, cur_value)

print(max_value)
print(min_value)
